rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,386 @@
+* If the user denies the request, the request will be updated to the denied state.
+
+#### Certificate retrieval
+
+After the headless login is initiated, the request will wait until local authentication is complete. This will be handled by the Teleport Proxy by using a resource watcher to wait until the `HeadlessAuthRequest` object is updated to the approved or denied state, or it expires.
+
+If the headless login request is approved, the Teleport Proxy will request MFA-verified, Single-use (1 minute TTL) user certificates from the Auth server with the `GenerateUserCerts` rpc. The proxy will provide the headless auth request ID and the request token previously generated to authorize the request. The login request details will be pulled from the `HeadlessAuthRequest` object to generate certificates.
+
+The resulting user certificates will then be sent to the client, and the client will complete the `tsh` request initially requested, e.g. `tsh ssh user@node01`.
+
+### Audit log
+
+The following actions will be tracked with audit events:
+
+* User initiates headless login
+* User approves/denies headless login request
+
+### Server changes
+
+Headless authentication has a unique API flow compared to other login methods.
+
+#### `POST /webapi/login/headless`
+
+This endpoint is used to initiate headless login. Like other login endpoints, this endpoint is not authenticated and can be called by anyone with access to the Teleport Proxy address.
+
+```go
+type HeadlessAuthRequest struct {
+    SSHLogin
+    // User is a teleport username
+    User string `json:"user"`
+    // RequestID is a uuid for the request
+    RequestID string
+    // Command is the client command being requested
+    Command string
+}
+
+// SSHLogin contains common SSH login parameters.
+type SSHLogin struct {
+    // ProxyAddr is the target proxy address
+    ProxyAddr string
+    // PubKey is SSH public key to sign
+    PubKey []byte
+    ...
+}
+
+// SSHLoginResponse is a user login response
+type SSHLoginResponse struct {
+    // Username contains the username for the login certificates
+    Username string `json:"username"`
+    // Cert is a PEM encoded SSH certificate signed by SSH certificate authority
+    Cert []byte `json:"cert"`
+    // TLSCertPEM is a PEM encoded TLS certificate signed by TLS certificate authority
+    TLSCert []byte `json:"tls_cert"`
+    // HostSigners is a list of signing host public keys trusted by proxy
+    HostSigners []TrustedCerts `json:"host_signers"`
+}
+```
+
+#### `rpc CreateHeadlessAuthRequest`
+
+This endpoint is used by the Teleport Proxy to create a new headless auth request.
+
+```proto
+service AuthService {
+  rpc CreateHeadlessAuthRequest(CreateHeadlessAuthRequestRequest) returns (HeadlessAuthRequest);
+}
+
+message CreateHeadlessAuthRequestRequest {
+  // RequestID is the headless login request uuid
+  string request_id = 1;
+}
+
+message HeadlessAuthRequest {
```

Comment:
Design suggestion: make "HeadlessAuthentication" into an entity, so you can have requests/responses that include it without having messages like "GetHeadlessAuthRequestRequest".

https://cloud.google.com/apis/design/resources